### PR TITLE
Merge unified trainer interface into a new nightly branch

### PR DESCRIPTION
**Pull request by listar2000**
## What is this PR about?

On a high-level, this PR will pin down a few high-level features in rLLM v0.3, including

1. A [`UnifiedTrainer`](https://github.com/rllm-org/rllm/blob/unified-trainer/rllm/experimental/unified_trainer.py) that controls the agentic RL training loop main logic, and each `backend` will implement the [`BackendProtocol`](https://github.com/rllm-org/rllm/blob/unified-trainer/rllm/experimental/protocol.py) interface.
2. A [`UnifiedWorkflowEngine`](https://github.com/rllm-org/rllm/blob/unified-trainer/rllm/experimental/engine/unified_workflow_engine.py) that becomes **backend-agnostic** -- i.e. implementing a standard 
  `execute_tasks(self, tasks: list[dict], task_ids: list[str]...) -> list[Episode]`
  function to generate rollouts of episodes.
3. A new advantage and policy loss calculation module ([`common/advantage.py`](https://github.com/rllm-org/rllm/blob/unified-trainer/rllm/experimental/common/advantage.py)) that attempts to standardize advantage estimator and RL algorithms across different `backends`.
4. A new configuration system that separates `rLLM`-specific parameters (e.g. configs for the training loop like `num_epochs`) away from backend-specific ones (e.g. the `actor_rollout_ref.xxx` for `Verl`). 

## Design docs
- [Discussions on episode -> trajectory group logic](https://www.notion.so/Shared-Episode-Transform-Protocol-2aa08daca1dd804396a3d74055be3b4b?source=copy_link)
- [Documentation on the `UnifiedTrainer`](https://docs.google.com/document/d/1Et3VSvZ7XdZv4y4e-Zprv-zp9e7N4_CXSuGuSppj1GU/edit?usp=sharing_)

## Incoming follow-up PRs
- @jeewoo-lee will implement the `SkyRL backend`.
- Someone will help implement more advantage estimators.
- @listar2000 will perform more testings and fixes (see below for known issues).
- @listar2000 will support training agent-env (i.e. `agent_ppo_trainer`) in the same trainer.

## Migration plan (TBD with separate PRs) into `main`

- [ ] Merge new utility function (e.g. better trajectory visualizer/logger and timer) into main.
- [ ] Merge the Tinker-related updates (including `TrajectoryGroup` data structure).
- [ ] Merge the Verl-related updates.
- [ ] Merge the `rLLM`-native advantage estimator and policy loss modules.
- [ ] Unify with other components of `rLLM`, e.g. `SDK`-based or `agent-env` trainers.

## Known issues to fix
- [x] Lack of `critic` metrics information when using `rLLM`-native advantage estimator.
- [ ] Training crash when turning on `broadcast` stepwise mode + working with `SimpleWorkflow`.
- [x] Remove the separate launchers and use the previous unified approach.

### CODE REVIEW

**Comment by listar2000** [comment]
@jeffreysijuntan Hi JT if you get time maybe review the `UnifiedTrainer` in `unified_trainer.py` and the `protocol.py` first (and no need to merge directly after reviewing). I will need to refactor the launcher a bit (should be quick) and attach some easily-reproducible script as the last step for this PR.
